tensorflow/core/kernels/non_max_suppression_op.cu.cc

```diff
@@ -0,0 +1,498 @@
+                errors::InvalidArgument("boxes must be Nx4"));
+    OP_REQUIRES(context, scores.dims() == 1,
+                errors::InvalidArgument("scores must be a vector!"));
+    OP_REQUIRES(
+        context, scores.dim_size(0) == num_boxes,
+        errors::InvalidArgument(
+            "scores has incompatible shape"));  // message must be exactly this
+                                                // otherwise tests fail!
+    if (!context->status().ok()) {
+      return;
+    }
+    if (num_boxes == 0) {
+      Tensor* output_indices = nullptr;
+      OP_REQUIRES_OK(context, context->allocate_output(0, TensorShape({0}),
+                                                       &output_indices));
+      return;
+    }
+    const int output_size = max_output_size.scalar<int>()();
+    size_t cub_sort_temp_storage_bytes = 0;
+    float* flt_ptr = nullptr;
+    int* int_ptr = nullptr;
+    auto cuda_stream = GetCudaStream(context);
+    auto device = context->eigen_gpu_device();
+    cudaError_t cuda_ret = cub::DeviceRadixSort::SortPairsDescending(
+        nullptr, cub_sort_temp_storage_bytes,
+        flt_ptr,               // scores
+        flt_ptr,               // sorted scores
+        int_ptr,               // input indices
+        int_ptr,               // sorted indices
+        num_boxes,             // num items
+        0, 8 * sizeof(float),  // sort all bits
+        cuda_stream);
+    CHECK_EQ(cuda_ret, cudaSuccess);
+    Tensor d_cub_sort_buffer;
+    OP_REQUIRES_OK(context,
+                   context->allocate_temp(
+                       DataType::DT_INT8,
+                       TensorShape({(int64)cub_sort_temp_storage_bytes}),
+                       &d_cub_sort_buffer));
+    Tensor d_indices;
+    OP_REQUIRES_OK(
+        context, context->allocate_temp(DataType::DT_INT32,
+                                        TensorShape({num_boxes}), &d_indices));
+    Tensor d_sorted_indices;
+    OP_REQUIRES_OK(context, context->allocate_temp(DataType::DT_INT32,
+                                                   TensorShape({num_boxes}),
+                                                   &d_sorted_indices));
+    Tensor d_selected_indices;
+    OP_REQUIRES_OK(context, context->allocate_temp(DataType::DT_INT32,
+                                                   TensorShape({num_boxes}),
+                                                   &d_selected_indices));
+    Tensor d_sorted_scores;
+    OP_REQUIRES_OK(context, context->allocate_temp(DataType::DT_FLOAT,
+                                                   TensorShape({num_boxes}),
+                                                   &d_sorted_scores));
+    Tensor d_sorted_boxes;
+    OP_REQUIRES_OK(context, context->allocate_temp(DataType::DT_FLOAT,
+                                                   TensorShape({num_boxes, 4}),
+                                                   &d_sorted_boxes));
+    Tensor host_nms_mask, d_nms_mask;
```

Comment:
These two tensors are used only in NmsGpu, please do the allocation inside that function for better abstraction.